scripts/test_hardware/test_esp32_api.py

```
@@ -0,0 +1,331 @@
+from typing import Optional
+
+from .result import TestResult, TestReport, Verdict
+
+
+DEFAULT_BASE_URL = "http://192.168.4.1"
+REQUEST_TIMEOUT = 5  # seconds
+
+
+def _get(url: str, timeout: float = REQUEST_TIMEOUT) -> tuple[int, str]:
+    """HTTP GET, returns (status_code, body)."""
+    req = urllib.request.Request(url)
```

Ruff (0.15.2)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)